### PR TITLE
rembadger dripper contract deployment

### PR DESCRIPTION
**Pull request by jorijnsmit**
issue #347 

wip

### CODE REVIEW

**Comment by jorijnsmit** [comment]
```
brownie run deployment/deploy_rembadger_dripper
```
or
```
brownie test tests/deployment/test_rembadger_dripper.py
```

File: contracts/RemBadgerDripper.sol [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    function setKeeper(address _newKeeper) public {
+        require(
+            _msgSender() == controller() || _msgSender() == governance(),
+            "RemBadgerDripper: onlyAuthorized"
+        );
+        _keeper = _newKeeper;
+    }
+
+    function released(address token) public view virtual override returns (uint256) {
```
Comment (GalloDaSballo) [inline contracts/RemBadgerDripper.sol:61]: No need to override this

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5c4b75a9cdda99af901b8c153d0d1e8a90b0139a/contracts/finance/VestingWallet.sol#L80

**Review by GalloDaSballo — COMMENTED** [review]
LGTM code wise

Basically a VestinWallet with extra role checks

Would like to see a vesting test with more formal math as you overridden the release function (although you just overrode roles to be fair)

File: contracts/RemBadgerDripper.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
+// contracts/RemBadgerDripper.sol
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/finance/VestingWallet.sol";
+
+contract RemBadgerDripper is VestingWallet {
+    mapping(address => uint256) private _erc20Released;
+    address private immutable _controller;
+    address private immutable _governance;
```
Comment (shuklaayush) [inline contracts/RemBadgerDripper.sol:10]:
```suggestion
    address public constant controller = 0x86cbD0ce0c087b482782c181dA8d191De18C8275;
    address public constant governance = 0xB65cef03b9B89f99517643226d76e286ee999e77;
```
You can keep the variables as public constants if they're going to be hardcoded. Constants are inlined during compilation and don't take a storage slot. Will save some `SLOAD` gas when the variables are accessed. The `public` automatically adds a getter.

File: contracts/RemBadgerDripper.sol [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    function controller() public view virtual returns (address) {
+        return _controller;
+    }
+
+    /**
+     * @dev Getter for the governance address.
+     */
+    function governance() public view virtual returns (address) {
+        return _governance;
+    }
```
Comment (shuklaayush) [inline contracts/RemBadgerDripper.sol:41]:
```suggestion

```

**Review by shuklaayush — CHANGES_REQUESTED** [review]
Some nits

File: contracts/RemBadgerDripper.sol [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    function release(address token) public virtual override {
+        require(
+            _msgSender() == keeper() || _msgSender() == controller() || _msgSender() == governance(),
+            "RemBadgerDripper: onlyAuthorized"
+        );
+        uint256 releasable = vestedAmount(token, uint64(block.timestamp)) - released(token);
+        _erc20Released[token] += releasable;
+        emit ERC20Released(token, releasable);
+        SafeERC20.safeTransfer(IERC20(token), beneficiary(), releasable);
```
Comment (shuklaayush) [inline contracts/RemBadgerDripper.sol:73]:
```suggestion
        VestingWallet.release(token);
```
This should also work

Comment (GalloDaSballo) [reply contracts/RemBadgerDripper.sol:73]: I agree with Aysuh, basically just add the check + use the super function

File: contracts/RemBadgerDripper.sol (hunk 2 above)
Comment (jorijnsmit) [reply contracts/RemBadgerDripper.sol:10]: i assume i can then also remove these lines from the constructor?
```
        _controller = 0x86cbD0ce0c087b482782c181dA8d191De18C8275; //techops
        _governance = 0xB65cef03b9B89f99517643226d76e286ee999e77; //dev
```

File: contracts/RemBadgerDripper.sol (hunk 4 above)
Comment (jorijnsmit) [reply contracts/RemBadgerDripper.sol]: thanks, this is what i was looking for, also solves having to overwrite/redefine `_erc20Released`!

File: contracts/RemBadgerDripper.sol (hunk 2 above)
Comment (shuklaayush) [reply contracts/RemBadgerDripper.sol]: Yep. Would also have to change references to `controller()`/`governance()` to `controller/governance`

**Comment by jorijnsmit** [comment]
deployed to rinkeby: https://rinkeby.etherscan.io/tx/0xed8b5157cdc49d603c58c761b9f36115a9e7c6c4633c3eed13526f8949ee7830

would like to do run some tests onchain:
- [ ] send funds
- [ ] call release before start
- [ ] setKeeper from gov
- [ ] setKeeper from random
- [ ] call release after start

**Comment by GalloDaSballo** [comment]
Code LG it's basicaly a VestedWallet with extra permission check

I'd like @dapp-whisperer to give review as well as this contract is not ugradeable

**Review by shuklaayush — APPROVED** [review]
LGTM. Can add upgradeability if desired

**Review by shuklaayush — APPROVED** [review]
LGTM

**Comment by GalloDaSballo** [comment]
I'd say after changes the code LGTM

OZ VestingWallet with extra sweep functionality

Saves from needing upgradeability

**Review by dapp-whisperer — APPROVED** [review]

**Pull request merged**